### PR TITLE
Migrate to Tokio 0.2

### LINKED ISSUE #485: Switch to std::future::Future
Transitioning from `futures` 0.1 to `std::future::Future` would greatly improve the user experience in the near future, especially with async/await stabilizing in Rust 1.39 and `std::future::Future` compatible versions of `tokio`, `hyper`, and `tower` on the horizon. It seems that `jsonrpc-core` is currently the only library in my personal JSON-RPC projects that still relies exclusively on `futures` 0.1, and it is blocking me from switching over to the new trait completely. I understand if this issue will sit on the backburner for the time being until at least `tokio` 0.2.0 is released, though.

### LINKED ISSUE #582: JSON Neo Tokio 0.2
tokio 0.2 has a lot of pertinent updates for IPC. #485 appears to still be in-progress which is obviously blocking, but it would be good to know if there are any additional requirements beyond that before jsonrpc-* can move to a ^0.2 version of tokio. Feel free to close this issue if the answer is "The moment #485 closes, tokio 0.2 is also ready" or "No plans to make that available, please move to jsonrpsee."

### PR DESCRIPTION
**Pull request by Xanewok**
Motivated to migrate RLS off Tokio 0.1 for IPC transport but since we're at it, it might make sense to migrate everything else in one fell swoop.

~This is a very early draft but opening it now to prevent duplicating work in case someone else starts working in it (or would like to help :grin:).~

Fixes #582
Fixes #485

### CODE REVIEW

**Comment by Xanewok** [comment]
This is ready for a preliminary review pass.

https://github.com/paritytech/jsonrpc/pull/603/commits/68e9be679782030efd2f55656eaefed66d7531c7 may be a red herring, I tried to fix hanging  `transports::http::tests::handles_connection_refused_error` test on Windows but I still can't get it to work there yet.

**Comment by Xanewok** [comment]
It looks like the Windows-specific bug is caused by TCP listener socket not being properly closed in `mio v0.6`. I rewrote the http server locally to use `hyper::Service` with graceful shutdown and confirmed that the associated event loop and the hyper server are closed as expected, yet test client still manages to get a TCP handshake and waits indefinitely after that.

The good news:
1. Manually dropping the TCP listener (created hackily via [`TcpListener::from_raw_socket`](https://doc.rust-lang.org/std/net/struct.TcpListener.html#method.from_raw_socket)) in the cleanup phase looks to be behaving as expected
2. Using freshly released hyper 0.14 (with freshly released Tokio 1.0 🎉 that includes mio 0.7 with Windows backend rewrite) seems to fix the issue as well, although I didn't inspect the difference in logic at the OS/low-level yet

My preference would be to pick the first option for now and focus on supporting Tokio 0.2 first; there are too many applications in the wild that are still using 0.1 and 0.2 (even we are transitioning just now!). After that, we can work on supporting Tokio 1.0.

**Comment by Xanewok** [comment]
Pushed a series of commits that remove our direct dependency on futures 0.1. I only tested the changes on Linux machine for now.

Before merging I'll squash and reformat to keep the CI happy :slightly_smiling_face:

**Comment by Xanewok** [comment]
Fixed macOS compilation issue in https://github.com/NikVolf/parity-tokio-ipc/pull/31/commits/0685cd6e272487ec9afb3455bcd865be118e64eb and force-pushed last commit to re-trigger macOS CI.

**Comment by Xanewok** [comment]
Pushed the final formatting commit, and so CI is green and the PR should be ready for review =)

File: stdio/src/lib.rs
Comment (tomusdrw) [inline stdio/src/lib.rs:77]: Instead of spawning the runtime, maybe we could just return a future here?

**Review by tomusdrw — COMMENTED** [review]
:tada: Great work! A couple of comments to answer, but overall this seems ready for merge.

File: server-utils/src/suspendable_stream.rs
Comment (Xanewok) [reply server-utils/src/suspendable_stream.rs:57]: You're right! I just mechanically adapted the code and tried to retain the semantics. I'll push a fix commit for this

Comment (Xanewok) [reply server-utils/src/suspendable_stream.rs]: Done in 7d4faa0a7002055645bd670d872ca919ccc3fbf0

File: http/src/lib.rs
Comment (Xanewok) [reply http/src/lib.rs:85]: Whenever possible, I just tried to prefer first-class `async` block support over some no-op wrappers from the external (staple, but still) `future` crate. I don't have strong feelings about that and can prefer using `future` wrappers for consistency whenever possible if you want

File: stdio/src/lib.rs
Comment (Xanewok) [reply stdio/src/lib.rs:77]: Sure; I just tried to adapt the existing code to work as before

Comment (Xanewok) [reply stdio/src/lib.rs]: Done in a883edcdde94764b0efb19d579982f60ed8a029a

File: core-client/transports/src/transports/ws.rs
Comment (Xanewok) [reply core-client/transports/src/transports/ws.rs:44]: Opened #605. I hope that's enough context; feel free to add more if necessary =)

File: Cargo.toml
Comment (Xanewok) [reply Cargo.toml:19]: Blocked on https://github.com/NikVolf/parity-tokio-ipc/pull/31 - that PR is fairly minimal; pinged @NikVolf already for review =)

File: core-client/transports/src/transports/http.rs
Comment (Xanewok) [reply core-client/transports/src/transports/http.rs]: It's not necessary (I did what previous code did); instead, we need to spawn a worker future on Tokio runtime directly.

Pushed dc1a81483c390625ffd712616f3219cd8c9f3d1b, let me know if that looks good

File: http/src/lib.rs
Comment (Xanewok) [reply http/src/lib.rs:638]: Done in #606

File: core-client/transports/src/transports/ws.rs
Comment (Xanewok) [reply core-client/transports/src/transports/ws.rs:108]: Right, so the way I wanted to structure it is that `poll_ready` is doing the heavy work and is responsible for draining our internal queue and gradually enqueueing the items in the underlying sink. As I understand, the previous behaviour was implemented to avoid greedily buffering until an explicit flush but rather to keep the queue as a fallback - if the sink is not yet ready, simply enqueue the item and process it later.

The more I think about it, the more I'm convinced we should just copy https://docs.rs/futures-util/0.3.8/src/futures_util/sink/buffer.rs.html :see_no_evil:

Comment (Xanewok) [reply core-client/transports/src/transports/ws.rs]: I think I made a mistake and shouldn't really do (pseudocode):
```rust
fn poll_$variant(...) {
    if !queue.is_empty() {
        self.poll_ready(cx)
    else {
        underlying.poll_$variant(cx)
    }
}
```
but rather
```rust
fn poll_$variant(...) {
    if !queue.is_empty() {
        match self.poll_ready(cx) {
            Poll::Pending => return Poll::Pending,
            err @ Poll::Ready(Err(..)) => return err,
            ok => underlying.poll_$variant(cx)
    }
    else {
        underlying.poll_$variant(cx)
    }
}
```
because there may be a case where we might call `poll_flush` with a non-empty queue, then drain it in delegated `poll_ready` call but ultimately not guarantee that the underlying send will be completed (which is the point of `poll_flush`, as you rightfully noticed), because we only call `poll_ready` on the underlying sink, in the end.

I used the modified (unbounded) version of `futures_util::sink::Buffer` version in 94a4b232feeb708d7e10c64320efd7609081524c, which already contains the aforementioned fix.

Let me know if this looks good now!

File: Cargo.toml
Comment (NikVolf) [reply Cargo.toml:19]: Published, sorry that missed that 👍

Comment (Xanewok) [reply Cargo.toml]: Thank you! Removed the patch from Cargo.toml and bumped parity-tokio-ipc to now-released v0.8

File: core-client/transports/src/transports/ws.rs
Comment (tomusdrw) [reply core-client/transports/src/transports/ws.rs]: Looks good! I'm pretty sure that when #605 is implemented we will get rid of this code anyway so no need to do any more sophisticated refactorings.

File: http/src/lib.rs
Comment (tomusdrw) [reply http/src/lib.rs:85]: Yeah, not using `future` crate is a pretty good argument, thanks!

File: core-client/transports/src/transports/ws.rs
Comment (tomusdrw) [inline core-client/transports/src/transports/ws.rs:140]: Can you elaborate on this conditional? Why can't we just call `try_empty_buffer` instead?

**Review by tomusdrw — APPROVED** [review]
lgtm!

File: core-client/transports/src/transports/ws.rs
Comment (Xanewok) [reply core-client/transports/src/transports/ws.rs:140]: I was motivated by the code in `futures_util` - I like how it's explicit in the two related `start_send`/`poll_ready` that it simply delegates to the underlying sink if the buffer queue is empty. You're right, however, that this is equivalent to just calling `try_empty_buffer`. Do you want me to change it?

Comment (tomusdrw) [reply core-client/transports/src/transports/ws.rs:140]: Nah, let's keep it like this, was just curious if I'm missing something.

**Comment by Xanewok** [comment]
It seems this only needs a second, separate approval and should be good to go :tada:

**Comment by niklasad1** [comment]
Yeah, I will review it, latest on Monday.

File: stdio/Cargo.toml
Comment (niklasad1) [inline stdio/Cargo.toml:20]: these `dev-dependecues` will leaked into the `release` build, however https://github.com/rust-lang/cargo/pull/8997 have been stabalized so should be ready soon I guess.

File: tcp/src/server.rs
Comment (niklasad1) [inline tcp/src/server.rs:1]:
```suggestion
```

**Review by niklasad1 — APPROVED** [review]
Nice work LGTM, I didn't review it super carefully except the Windows mio hacks.

**Comment by Xanewok** [comment]
While going through the diff again I noticed that I removed a Tokio re-export in ipc server - rolled that back in https://github.com/paritytech/jsonrpc/pull/603/commits/a05cd4ed8ae48b3838831eecb0dd608fac13de50 by adding the re-export.

Since this is all green, I'm merging it. Thank you for taking your time to review this!

**Pull request merged**
